### PR TITLE
Code Evaluator refactoring

### PR DESCRIPTION
**Pull request by ankitjavalkar**

### CODE REVIEW

File: yaksh/bash_code_evaluator.py [hunk 1]
```diff
@@ -9,26 +9,31 @@
...
+from .base_evaluator import BaseEvaluator
 from .file_utils import copy_files, delete_files
 
 
-class BashCodeEvaluator(CodeEvaluator):
+class BashCodeEvaluator(BaseEvaluator):
     # Private Protocol ##########
-    def setup(self):
-        super(BashCodeEvaluator, self).setup()
+    def __init__(self, metadata, test_case_data):
```
Comment (prabhuramachandran) [inline yaksh/bash_code_evaluator.py]: In the next refactor you could actually refactor this to use all keyword args like `__init__(self, user_answer, file_paths, ...)` and use **metadata when instantiating it?

File: yaksh/language_registry.py [hunk 2]
```diff
@@ -14,15 +14,16 @@ def get_registry():
         registry = _LanguageRegistry()
     return registry
 
-def unpack_json(json_data):
-    data = json.loads(json_data)
-    return data
+# def unpack_json_to_python_obj(json_data):
```
Comment (prabhuramachandran) [inline yaksh/language_registry.py]: Remove commented code.

File: yaksh/settings.py [hunk 3]
```diff
@@ -19,6 +19,26 @@
 # host.org/foo/exam set URL_ROOT='/foo'
 URL_ROOT = ''
 
+# code_evaluators = {
```
Comment (prabhuramachandran) [inline yaksh/settings.py]: Remove commented code.

File: yaksh/code_evaluator.py [hunk 4]
```diff
@@ -61,9 +74,10 @@ def __init__(self, in_dir=None):
         msg = 'Code took more than %s seconds to run. You probably '\
               'have an infinite loop in your code.' % SERVER_TIMEOUT
         self.timeout_msg = msg
-        self.in_dir = in_dir
+        self.in_dir = in_dir if in_dir else MY_DIR
```
Comment (prabhuramachandran) [inline yaksh/code_evaluator.py]: Maybe the main `CodeEvaluator` should renamed to `Grader` or something like that.  Otherwise there are too many "Evaluators", there is `CodeEvaluator` and `BaseEvaluator` which are not related via inheritance which is confusing.

File: yaksh/code_evaluator.py [hunk 5]
```diff
@@ -97,45 +113,55 @@ def setup(self):
         if self.in_dir:
             if not os.path.exists(self.in_dir):
                 os.makedirs(self.in_dir)
-        self._change_dir(self.in_dir)
 
-    def safe_evaluate(self, user_answer, partial_grading, test_case_data, file_paths=None):
+    def get_evaluator_objects(self, kwargs):
+        metadata = kwargs.get('metadata')
+        test_case_data = kwargs.get('test_case_data')
```
Comment (prabhuramachandran) [inline yaksh/code_evaluator.py]: Why are you keeping two different things, metadata and test_case_data?  Why not just have them all as kwargs and send them as such?  What does this split bring you?

File: yaksh/code_evaluator.py [hunk 6]
```diff
@@ -86,7 +100,9 @@ def evaluate(self, **kwargs):
         """
 
         self.setup()
-        success, error, weight = self.safe_evaluate(**kwargs)
+        test_case_instances = self.get_evaluator_objects(kwargs)
+        with change_dir(self.in_dir):
+            success, error, weight = self.safe_evaluate(test_case_instances)
         self.teardown()
 
         result = {'success': success, 'error': error, 'weight': weight}
```
Comment (prabhuramachandran) [inline yaksh/code_evaluator.py]: Should 'weight' really be 'grade' or 'mark_fraction' or 'mark'?

File: yaksh/code_server.py [hunk 7]
```diff
@@ -84,19 +85,19 @@ def check_code(self, language, test_case_type, json_data, in_dir=None):
...
         """
-        code_evaluator = create_evaluator_instance(language,
-            test_case_type,
-            json_data,
-            in_dir
-        )
-        data = unpack_json(json_data)
-        result = code_evaluator.evaluate(**data)
+        data = self.unpack_json_to_python_obj(json_data)
+        code_eval_instance = CodeEvaluator(in_dir)
```
Comment (prabhuramachandran) [inline yaksh/code_server.py]: `code_eval = CodeEvaluator(in_dir)` or better
`grader = Grader(in_dir)`

File: yaksh/language_registry.py [hunk 8]
```diff
@@ -14,15 +14,16 @@ def get_registry():
...
+#     data = json.loads(json_data)
+#     return data
 
-def create_evaluator_instance(language, test_case_type, json_data, in_dir):
+def create_evaluator_instance(metadata, test_case): #create_evaluator_instance
```
Comment (prabhuramachandran) [inline yaksh/language_registry.py]: Whats the comment doing here?

File: yaksh/code_server.py [hunk 9]
```diff
@@ -84,19 +85,19 @@ def check_code(self, language, test_case_type, json_data, in_dir=None):
...
+        result = code_eval_instance.evaluate(data) #language, test_case_type, 
 
         # Put us back into the server pool queue since we are free now.
         self.queue.put(self.port)
 
         return json.dumps(result)
 
+    def unpack_json_to_python_obj(self, json_data):
```
Comment (prabhuramachandran) [inline yaksh/code_server.py]: This method seems unnecessary, its a single line.

File: yaksh/evaluator_tests/test_scilab_evaluation.py [hunk 10]
```diff
@@ -4,77 +4,106 @@
...
         tmp_in_dir_path = tempfile.mkdtemp()
         self.test_case_data = [{"test_case": "scilab_files/test_add.sce",
+                                "test_case_type": "standardtestcase",
                                 "weight": 0.0
                                 }]
         self.in_dir = tmp_in_dir_path
         self.timeout_msg = ("Code took more than {0} seconds to run. "
                             "You probably have an infinite loop" 
                             " in your code.").format(SERVER_TIMEOUT)
+        code_evaluator.SERVER_TIMEOUT = 9
```
Comment (prabhuramachandran) [inline yaksh/evaluator_tests/test_scilab_evaluation.py]: Why is this added?

File: yaksh/evaluator_tests/test_scilab_evaluation.py [hunk 11]
```diff
@@ -4,77 +4,106 @@
...
         self.file_paths = None
 
     def tearDown(self):
+        code_evaluator.SERVER_TIMEOUT = 4
```
Comment (prabhuramachandran) [inline yaksh/evaluator_tests/test_scilab_evaluation.py]: Why is this done and hard-coded to 4?

File: yaksh/evaluator_tests/test_scilab_evaluation.py [hunk 12]
```diff
@@ -4,77 +4,106 @@
...
+        evaluator = CodeEvaluator(self.in_dir)
+        result = evaluator.evaluate(kwargs)
+
         lines_of_error = len(result.get('error').splitlines())
         self.assertFalse(result.get('success'))
         self.assertIn("Message", result.get('error'))
         self.assertTrue(lines_of_error > 1)
 
     def test_infinite_loop(self):
+        code_evaluator.SERVER_TIMEOUT = 4
```
Comment (prabhuramachandran) [inline yaksh/evaluator_tests/test_scilab_evaluation.py]: Why?

File: yaksh/models.py [hunk 13]
```diff
@@ -1138,12 +1154,13 @@ def __str__(self):
 
 
 class StdioBasedTestCase(TestCase):
```
Comment (prabhuramachandran) [inline yaksh/models.py]: Should this be `StdIOBasedTestCase`?

File: yaksh/python_stdio_evaluator.py [hunk 14]
```diff
@@ -28,27 +28,33 @@ def redirect_stdout():
         sys.stdout = old_target  # restore to the previous value
 
 
-class PythonStdioEvaluator(CodeEvaluator):
+class PythonStdioEvaluator(BaseEvaluator):
```
Comment (prabhuramachandran) [inline yaksh/python_stdio_evaluator.py]: `PythonStdIOEvaluator`?

File: yaksh/settings.py (hunk 3 above)
Comment (ankitjavalkar) [inline yaksh/settings.py]: Done

File: yaksh/language_registry.py (hunk 8 above)
Comment (ankitjavalkar) [inline yaksh/language_registry.py]: Done

File: yaksh/language_registry.py (hunk 2 above)
Comment (ankitjavalkar) [inline yaksh/language_registry.py]: Done

File: yaksh/code_evaluator.py (hunk 6 above)
Comment (ankitjavalkar) [inline yaksh/code_evaluator.py]: Done

File: yaksh/code_server.py (hunk 7 above)
Comment (ankitjavalkar) [inline yaksh/code_server.py]: Done

File: yaksh/code_server.py (hunk 9 above)
Comment (ankitjavalkar) [inline yaksh/code_server.py]: Done

File: yaksh/python_stdio_evaluator.py (hunk 14 above)
Comment (ankitjavalkar) [inline yaksh/python_stdio_evaluator.py]: Done

File: yaksh/evaluator_tests/test_scilab_evaluation.py (hunk 10 above)
Comment (ankitjavalkar) [inline yaksh/evaluator_tests/test_scilab_evaluation.py]: Generally Java & Scilab testcase do not pass in the default 4 sec time frame. So I over-rode the timeout and increased it.

File: yaksh/evaluator_tests/test_scilab_evaluation.py (hunk 12 above)
Comment (ankitjavalkar) [inline yaksh/evaluator_tests/test_scilab_evaluation.py]: Sorry, this should not be here. Fixed

File: yaksh/bash_code_evaluator.py (hunk 1 above)
Comment (ankitjavalkar) [inline yaksh/bash_code_evaluator.py]: Issue created. But I will have this fixed asap

File: yaksh/code_evaluator.py (hunk 4 above)
Comment (ankitjavalkar) [inline yaksh/code_evaluator.py]: Done

File: yaksh/code_evaluator.py (hunk 5 above)
Comment (ankitjavalkar) [inline yaksh/code_evaluator.py]: I had earlier assumed that we may need to pop some data or may have to pass the arguments to different classes/methods (metadata is mostly question/user answer related and test case data is test case specific)

I guess I also assumed that this would be more readable and segregated.

File: yaksh/bash_code_evaluator.py (hunk 1 above)
Comment (ankitjavalkar) [inline yaksh/bash_code_evaluator.py]: Issue #164

File: yaksh/code_evaluator.py (hunk 5 above)
Comment (ankitjavalkar) [inline yaksh/code_evaluator.py]: Issue #165
